### PR TITLE
Histogram API Improvements

### PR DESCRIPTION
**Pull request by jamesmcclain**
Numerous changes have been made to the Histogram API.  Most notably from an external perspective, the "get" and "generate" verbs have been removed from various methods, unused histogram implementations have gone, the `toJSON` method has gone, and the `merge` method has been added.

Also, "double" versions (to accompany "int" versions) of various `geotrellis.raster` and `geotrellis.spark` methods that make use of histograms have been created.  Notably, the `StrictColorClassifier.fromQuantileBreaks` function can now produce color breaks using the double histogram implementation.  Tests have also been created for those.

Various functions/methods in the png rendering code, jpeg rendering code, and `geotrellis.engine` project which use histograms have not yet been given "double" versions.

### CODE REVIEW

File: raster/src/main/scala/geotrellis/raster/histogram/ArrayHistogram.scala [hunk 1]
```diff
@@ -79,10 +75,10 @@ class ArrayHistogram(val counts: Array[Int], var total: Int)
     counts(i) += count
   }
 
-  def getItemCount(i: Int) = counts(i)
+  def itemCount(i: Int) = counts(i)
 
   // REFACTOR: use Option
```
Comment (lossyrob) [inline raster/src/main/scala/geotrellis/raster/histogram/ArrayHistogram.scala]: We should remove these REFACTOR lines, unless what they are trying to do make sense, then we should do them and remove them.

File: raster/src/main/scala/geotrellis/raster/histogram/ArrayHistogram.scala [hunk 2]
```diff
@@ -101,4 +97,11 @@ class ArrayHistogram(val counts: Array[Int], var total: Int)
...
     return Int.MinValue
   }
+
+  def merge(histogram: Histogram[Int]): Histogram[Int] = {
+    val total = ArrayHistogram(histogram.bucketCount())
+
+    total.update(this); total.update(histogram)
+    total
+  }
 }
```
Comment (lossyrob) [inline raster/src/main/scala/geotrellis/raster/histogram/ArrayHistogram.scala]: I see you kept ArrayHistogram around...it was on the chopping block, what was the justification for keeping it in?

File: raster/src/main/scala/geotrellis/raster/histogram/StreamingHistogram.scala [hunk 3]
```diff
@@ -396,25 +437,29 @@ class StreamingHistogram(
...
   }
 
-  def getPercentile(q: Double): Double =
-    getPercentileBreaks(List(q)).head
+  def percentile(q: Double): Double =
+    percentileBreaks(List(q)).head
 
-  def getQuantileBreaks(num: Int): Array[Double] =
-    getPercentileBreaks(List.range(0,num).map(_ / num.toDouble)).toArray
+  def quantileBreaks(num: Int): Array[Double] =
```
Comment (lossyrob) [inline raster/src/main/scala/geotrellis/raster/histogram/StreamingHistogram.scala]: Can we put some scaladocs that talk about the guarantee that we talked about this morning, that there will be no item in the value set this histogram represents that is outside of the bounds of the quantile breaks returned?

File: raster/src/main/scala/geotrellis/raster/render/color/ColorClassifier.scala [hunk 4]
```diff
@@ -429,9 +429,11 @@ object StrictColorClassifier {
     colorClassifier
   }
 
-  def fromQuantileBreaks(histogram: Histogram[Int], colors: Array[RGBA]) = {
-    val breaks = histogram.getQuantileBreaks(colors.length)
-    apply(breaks zip colors)
+  def fromQuantileBreaks[X <: AnyVal : ClassTag](histogram: Histogram[X], colors: Array[RGBA]) = {
+    histogram.quantileBreaks(colors.length) match {
+      case breaks: Array[Int] => apply(breaks zip colors)
+      case breaks: Array[Double] => apply(breaks zip colors)
```
Comment (lossyrob) [inline raster/src/main/scala/geotrellis/raster/render/color/ColorClassifier.scala]: Need to handle the default case here.

File: raster/src/main/scala/geotrellis/raster/render/color/ColorClassifier.scala [hunk 5]
```diff
@@ -429,9 +429,11 @@ object StrictColorClassifier {
     colorClassifier
   }
 
-  def fromQuantileBreaks(histogram: Histogram[Int], colors: Array[RGBA]) = {
-    val breaks = histogram.getQuantileBreaks(colors.length)
-    apply(breaks zip colors)
+  def fromQuantileBreaks[X <: AnyVal : ClassTag](histogram: Histogram[X], colors: Array[RGBA]) = {
```
Comment (lossyrob) [inline raster/src/main/scala/geotrellis/raster/render/color/ColorClassifier.scala]: Not into this generic function that just checks two types.

I think if there were simply an overload that took `Array[Double], Array[RGBA]` (or traversables in general), and did the zip for you, that would be sufficient.

File: spark/src/main/scala/geotrellis/spark/mapalgebra/zonal/ZonalTileRDDMethods.scala [hunk 6]
```diff
@@ -14,12 +14,12 @@ import spire.syntax.cfor._
 
 trait ZonalTileRDDMethods[K] extends TileRDDMethods[K] {
 
-  private def mergeMaps(a: Map[Int, Histogram[Int]], b: Map[Int, Histogram[Int]]) = {
+  private def mergeMaps[X <: AnyVal](a: Map[Int, Histogram[X]], b: Map[Int, Histogram[X]]) = {
```
Comment (lossyrob) [inline spark/src/main/scala/geotrellis/spark/mapalgebra/zonal/ZonalTileRDDMethods.scala]: Why X? `T` seems more appropriate here.

File: raster/src/main/scala/geotrellis/raster/histogram/ArrayHistogram.scala (hunk 2 above)
Comment (jamesmcclain) [inline raster/src/main/scala/geotrellis/raster/histogram/ArrayHistogram.scala]: Evidently I overlooked it at the beginning of the list.

File: raster/src/main/scala/geotrellis/raster/histogram/ArrayHistogram.scala (hunk 1 above)
Comment (jamesmcclain) [inline raster/src/main/scala/geotrellis/raster/histogram/ArrayHistogram.scala]: Okay, will-do.

File: raster/src/main/scala/geotrellis/raster/histogram/StreamingHistogram.scala (hunk 3 above)
Comment (jamesmcclain) [inline raster/src/main/scala/geotrellis/raster/histogram/StreamingHistogram.scala]: Will-do.

File: spark/src/main/scala/geotrellis/spark/mapalgebra/zonal/ZonalTileRDDMethods.scala (hunk 6 above)
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/mapalgebra/zonal/ZonalTileRDDMethods.scala]: It shall be changed.

**Comment by jamesmcclain** [comment]
I believe that all comments prior to this one have been addressed.  The refactor suggested in the comments has been made, as well as one logically implied by it.  The respective commits can be removed if those changes are not desired.

**Pull request merged**